### PR TITLE
[Rust Client] several small fixes to code generation errors

### LINKED ISSUE #8380: [BUG][Rust Client] Duplicated crate::models in Params struct
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description
Generated rust client has duplicated `crate::models` string in enum field types. But rust doesn't like it and fails with `'crate' in paths can only be used in start position` in compilation phase. Also tested with `a` as query parameter, same results

##### openapi-generator version
5.0.0
5.0.1-SNAPSHOT
6.0.0-SNAPSHOT

##### OpenAPI declaration file content or url
```yaml
---
openapi: 3.0.1
info:
  title: v3
  version: 3.0.0
servers:
  - url: https://example.com
paths:
  /v3/{a}:
    get:
      parameters:
        - name: a
          in: path
          required: true
          schema:
            $ref: '#/components/schemas/ReleaseType'
      responses:
        "200":
          description: search results matching criteria
components:
  schemas:
    ReleaseType:
      default: ga
      enum:
        - ga
        - ea
      type: string
```

##### Generation Details
Generator: `rust`
Config:
```json
{
  "library": "reqwest",
  "packageName": "api",
  "packageVersion": "1.0.0",
  "supportAsync": "true",
  "useSingleRequestParameter": "true"
}
```
##### Expected output (`src/apis/default_api.rs`)
```rust
/// struct for passing parameters to the method `v3_a_get`
#[derive(Clone, Debug)]
pub struct V3AGetParams {
    pub a: crate::models::ReleaseType
}
```

##### Actual output (`src/apis/default_api.rs`)
```rust
/// struct for passing parameters to the method `v3_a_get`
#[derive(Clone, Debug)]
pub struct V3AGetParams {
    pub a: crate::models::crate::models::ReleaseType
}
```

### PR DESCRIPTION
**Pull request by ahl**
This fixes a few rust code generation issues. I've included links to small test cases.

- compile errors incorrectly referencing types e.g. `crate::models::crate::models::...`
- compile errors on types named `File`
- compile errors on cyclic types
- compile warnings on incorrectly named enum values

Closes #8380 

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request. (hi @frol @farcaller @richardwhiuk @paladinzh)

### CODE REVIEW

**Comment by ahl** [comment]
See these simple test cases:
- https://github.com/ahl/openapi-tests/blob/master/src/cycle_types.json
- https://github.com/ahl/openapi-tests/blob/master/src/enum_names.json
- https://github.com/ahl/openapi-tests/blob/master/src/file_type.json

Each includes the errors currently produced by master

**Comment by antifuchs** [comment]
This PR does get much further than the latest release in generating the client that I'm interested in ([docspell's](https://docspell.org/openapi/docspell-openapi.yml)). Generating from the linked .yml file results in these `cargo check` errors:

```
    Checking openapi v1.0.0 (/Users/asf/Hacks/docspell-api-client/docspell-openapi)
error: named argument never used
   --> src/apis/item_multi_edit_api.rs:366:103
    |
366 |     let local_var_uri_str = format!("{}/sec/items/customfieldremove", configuration.base_path, itemId=crate::apis::urlencode(item_id));
    |                                     -------------------------------- formatting specifier missing     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ named argument never used

error[E0412]: cannot find type `ItemUploadMeta` in module `crate::models`
   --> src/apis/integration_endpoint_api.rs:106:128
    |
106 | ...ation, id: &str, meta: Option<crate::models::ItemUploadMeta>, file: Option<Vec<std::path::PathBuf>>) -> Result<crate::models::BasicRes...
    |                                                 ^^^^^^^^^^^^^^ not found in `crate::models`

error[E0412]: cannot find type `OneOfReferenceListOrganizationList` in module `crate::models`
  --> src/apis/organization_api.rs:55:143
   |
55 | ...on<&str>) -> Result<crate::models::OneOfReferenceListOrganizationList, Error<SecOrganizationGetError>> {
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `crate::models`

error[E0412]: cannot find type `OneOfReferenceListPersonList` in module `crate::models`
  --> src/apis/person_api.rs:55:137
   |
55 | ...ption<&str>) -> Result<crate::models::OneOfReferenceListPersonList, Error<SecPersonGetError>> {
   |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `crate::models`

error[E0412]: cannot find type `ItemUploadMeta` in module `crate::models`
  --> src/apis/upload_api.rs:89:123
   |
89 | ...ation, id: &str, meta: Option<crate::models::ItemUploadMeta>, file: Option<Vec<std::path::PathBuf>>) -> Result<crate::models::BasicRes...
   |                                                 ^^^^^^^^^^^^^^ not found in `crate::models`

error[E0412]: cannot find type `ItemUploadMeta` in module `crate::models`
   --> src/apis/upload_api.rs:122:146
    |
122 | ..., item_id: &str, meta: Option<crate::models::ItemUploadMeta>, file: Option<Vec<std::path::PathBuf>>) -> Result<crate::models::BasicRes...
    |                                                 ^^^^^^^^^^^^^^ not found in `crate::models`

error[E0412]: cannot find type `ItemUploadMeta` in module `crate::models`
   --> src/apis/upload_api.rs:190:127
    |
190 | ..., item_id: &str, meta: Option<crate::models::ItemUploadMeta>, file: Option<Vec<std::path::PathBuf>>) -> Result<crate::models::BasicRes...
    |                                                 ^^^^^^^^^^^^^^ not found in `crate::models`

error[E0412]: cannot find type `ItemUploadMeta` in module `crate::models`
   --> src/apis/upload_api.rs:231:109
    |
231 | ...::Configuration, meta: Option<crate::models::ItemUploadMeta>, file: Option<Vec<std::path::PathBuf>>) -> Result<crate::models::BasicRes...
    |                                                 ^^^^^^^^^^^^^^ not found in `crate::models`

error[E0283]: type annotations needed
    --> src/apis/organization_api.rs:87:9
     |
87   |         serde_json::from_str(&local_var_content).map_err(Error::from)
     |         ^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `T` declared on the function `from_str`
     |
    ::: /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/serde_json-1.0.63/src/de.rs:2584:8
     |
2584 |     T: de::Deserialize<'a>,
     |        ------------------- required by this bound in `serde_json::from_str`
     |
     = note: cannot satisfy `_: Deserialize<'_>`
help: consider specifying the type arguments in the method call
     |
87   |         serde_json::from_str(&local_var_content).map_err::<F, O>(Error::from)
     |                                                         ^^^^^^^^
help: consider specifying the type arguments in the function call
     |
87   |         serde_json::from_str::<'a, T>(&local_var_content).map_err(Error::from)
     |                             ^^^^^^^^^

error[E0283]: type annotations needed
    --> src/apis/person_api.rs:87:9
     |
87   |         serde_json::from_str(&local_var_content).map_err(Error::from)
     |         ^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `T` declared on the function `from_str`
     |
    ::: /Users/asf/.cargo/registry/src/github.com-1ecc6299db9ec823/serde_json-1.0.63/src/de.rs:2584:8
     |
2584 |     T: de::Deserialize<'a>,
     |        ------------------- required by this bound in `serde_json::from_str`
     |
     = note: cannot satisfy `_: Deserialize<'_>`
help: consider specifying the type arguments in the method call
     |
87   |         serde_json::from_str(&local_var_content).map_err::<F, O>(Error::from)
     |                                                         ^^^^^^^^
help: consider specifying the type arguments in the function call
     |
87   |         serde_json::from_str::<'a, T>(&local_var_content).map_err(Error::from)
     |                             ^^^^^^^^^
```

**Comment by wing328** [comment]
cc @frol (2017/07) @farcaller (2017/08) @richardwhiuk (2019/07) @paladinzh (2020/05)

**Comment by ahl** [comment]
Hey @wing328 -- how can I help to move this forward? Thanks.

**Comment by wing328** [comment]
@ahl did you see the issues reported by @antifuchs when he tried your PR/fix? Are you able to reproduce those locally?

**Comment by ahl** [comment]
> @ahl did you see the issues reported by @antifuchs when he tried your PR/fix? Are you able to reproduce those locally?

Yes, he said "This PR does get much further than the latest release in generating the client that I'm interested in..." i.e. that it's a significant improvement over the latest version.

The Rust generator struggles with quite a few constructions--too many for me to fix in this PR. With over 2k open issues, it's a little hard to understand what's filed or not, but from the output @antifuchs posted I recognize several issues that exist independent of this PR:

- the ItemUploadMeta error is due to incorrect handling of multipart/form-data
- no support for oneOf schemas (hence the several
- general sloppiness and poor test coverage

**Comment by wing328** [comment]
> Yes, he said "This PR does get much further than the latest release in generating the client that I'm interested in..." i.e. that it's a significant improvement over the latest version.

Totally agree on that. My question is wondering if you've looked into the issue he reported and the finding, and it's perfectly fine to have another PR to fix those instead of squeezing all the fixes into one giant PR.

> no support for oneOf schemas (hence the several

We are definitely looking for contributors to help on that as we've already come up with the implementation in some other generators (java, csharp, go, python, and more). We can use these as the reference implementations to start with.

I'll take another look over the weekend and merge if no further feeddback/question from me.

Thanks again for the PR. Have a nice weekend.

**Comment by ahl** [comment]
Thanks, William. Yes I was able to reproduce those. I have some ideas about how to fix them, but this code feels quite fragile. I'm unsure about test coverage and wouldn't be confident that a fix in one place wouldn't regress another. In this PR I tried to stick to some pretty clear wins like resolving a fundamental issue with circular types.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java [hunk 1]
```diff
@@ -224,7 +224,11 @@ public RustClientCodegen() {
                         Map<String, Object> mas = new HashMap<>();
                         mas.put("modelName", camelize(mappedModel.getModelName()));
                         mas.put("mappingName", mappedModel.getMappingName());
-                        List<CodegenProperty> vars = model.getVars();
+                        
+                        // TODO: deleting the variable from the array was
+                        // problematic; I don't know what this is supposed to do
+                        // so I'm just cloning it for the moment
+                        List<CodegenProperty> vars = new ArrayList<>(model.getVars());
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:231]: @ahl when you've time, can you please elaborate on the problem ?

Comment (ahl) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:231]: Yeah, sorry to leave that turd. This code section seems to be handling discriminators and relocates a particular variable into the discriminators variables. It previously did so destructively which caused the original variable to be absent.

This is definitely ugly and poorly structured, but I don't have a good sense of how it should be structured or how to properly and exhaustively test some of the subtle semantics here.

Comment (wing328) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:231]: No worries bro. Do you have a spec that can reproduce the issue that you encountered with this line/section of code?

**Comment by wing328** [comment]
@antifuchs I wonder if you can open an issue with the details to track the problem you reported above. Thanks.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java (hunk 1 above)
Comment (ahl) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:231]: I was running the influxdb openapi spec through it: https://docs.influxdata.com/influxdb/v2.0/api/

**Comment by wing328** [comment]
@thanks for the PR, which has been included in the v5.1.0 release: https://twitter.com/oas_generator/status/1373636654024380423

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java (hunk 1 above)
Comment (wing328) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:231]: OK. Will take a look when I can find the time later.

**Pull request merged**